### PR TITLE
update to R 4.4.0

### PR DESCRIPTION
**Pull request by alexvpickering**
# Description
- updates R to 4.4.0 and all packages
- updates code to reflect changes to Seurat (v5)

Things tested:
- [x] Default pipeline with 10x dataset
- [x] "within" and "between" DE with 10x and Seurat datasets
- [x] All integration methods with 10x dataset

staging:

https://ui-alex-worker357-pipeline363.staging.single-cell-platform.net

# Details
#### URL to issue
N/A

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/pipeline/pull/363

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: r/R/cell_annotation_sctype.R [hunk 1]
```diff
@@ -74,13 +72,13 @@ ScTypeAnnotate <- function(req, data) {
 #' @examples
 get_formatted_data <- function(scdata, active_assay) {
 
-  scale_data <- scdata[[active_assay]]@scale.data
+  scale_data <- scdata[[active_assay]]$scale.data
```
Comment (gerbeldo) [inline r/R/cell_annotation_sctype.R:75]: here's what I meant in the pipeline question. do old seurat objects work with these changes? I assume they do, but just in case 😬

File: r/R/utilities_expression.R [hunk 2]
```diff
@@ -113,7 +119,7 @@ completeExpression <- function(expression, cell_ids) {
     on = .(cell_ids)
     ]
 
-  expression[, cell_ids := NULL]
```
Comment (gerbeldo) [inline r/R/utilities_expression.R:116]: why this change? in-place deletion is much much faster. And all this expression code needs to be as fast as it can be (data exploration user experience currently depends heavily on performance here, in particular to plot the gene expression values in the umap.)

File: r/R/utilities_expression.R [hunk 3]
```diff
@@ -100,8 +107,7 @@ completeExpression <- function(expression, cell_ids) {
     return(expression)
   }
 
-  expression[, cell_ids := cell_ids]
```
Comment (gerbeldo) [inline r/R/utilities_expression.R:103]: did you encounter any problems with in-place assignment here? I'm asking because I ran into some issues with the expression code a while ago.

File: r/R/utilities_expression.R [hunk 4]
```diff
@@ -203,7 +209,7 @@ getStats <- function(data) {
 sparsify <- function(expression) {
   data.table::setnafill(expression, fill = 0)
   sparse_matrix <-
-    Matrix::Matrix(Matrix::as.matrix(expression), sparse = T)
+    as(as.matrix(expression), 'dgCMatrix')
```
Comment (gerbeldo) [inline r/R/utilities_expression.R:212]: yesss that conversion was horrible.

File: r/Dockerfile [hunk 5]
```diff
@@ -19,7 +19,7 @@ RUN echo ".libPaths(c('$RENV_LIB', .libPaths()))" >> $(R RHOME)/etc/Rprofile.sit
 
 # install renv to install required R packages
 RUN R -q -e "install.packages('remotes', repos = c(CRAN = 'https://cloud.r-project.org'))" && \
-    R -q -e "remotes::install_github('rstudio/renv@0.16.0')" && \
+    R -q -e "remotes::install_github('rstudio/renv@v1.0.7')" && \
```
Comment (gerbeldo) [inline r/Dockerfile:22]: same suggestion, changing to PPM cran repo to speedup build.

**Review by gerbeldo — APPROVED** [review]
looks good!

File: r/Dockerfile (hunk 5 above)
Comment (alexvpickering) [reply r/Dockerfile:22]: single package/dependencies installation probably not worthwhile

File: r/R/cell_annotation_sctype.R (hunk 1 above)
Comment (alexvpickering) [reply r/R/cell_annotation_sctype.R:75]: Yes they do. Seurat V5 and V3 objects can be accessed in this manner (`scdata[[assay]]$layer`) while only Seurat V3 objects can be accessed in the old manner (`scdata[[assay]]@slot`)

File: r/R/utilities_expression.R (hunk 3 above)
Comment (alexvpickering) [reply r/R/utilities_expression.R:103]: Yes, there was some side effects outside of the function (something in the parent function was being updated unintentionally). 

This kind of operation in incredibly fast in base R -- no way that it limits the speed of this function at all.

File: r/R/utilities_expression.R (hunk 2 above)
Comment (alexvpickering) [reply r/R/utilities_expression.R:116]: Again, due to unintended side effects in parent function and because removal of a column is incredible fast (I would guess on order of nanoseconds). This is premature optimization -- other things will be slower in this function and so optimizing this will have no effect.

File: r/R/utilities_expression.R (hunk 4 above)
Comment (alexvpickering) [reply r/R/utilities_expression.R:212]: Changed this because some of the tests were creating a `dTiMatrix` or something like that which led to errors. This just makes it explicit the type of sparse matrix we want.

Comment (gerbeldo) [reply r/R/utilities_expression.R:212]: yep, newer matrix converts (mathematically) diagonal matrices to that diagonal matrix type. The other solution is to add `diagonal = F` to the Matrix::Matrix call, but this is much nicer.

Comment (alexvpickering) [reply r/R/utilities_expression.R:212]: as okay cool -- thanks for the heads up. We do have a few other calls of `Matrix::Matrix(..., sparse = T)`. They don't seem to be causing any errors in our tests but maybe they should be updated as well to prevent `dTiMatrix` shenanigans
